pose_prediction/const_vel/plugin.cpp

```diff
@@ -0,0 +1,410 @@
+    mutable Eigen::Quaternionf offset {Eigen::Quaternionf::Identity()};
+    mutable std::shared_mutex offset_mutex;
+
+    // Based on https://github.com/rpng/open_vins/issues/92
+    // Returns a pair of the predictor state and the time associated with the
+    // most recent imu reading used to perform this prediction.
+    std::pair<Eigen::Matrix<double, 13, 1>, time_point> predict_const_vel(double dt) const {
+        const auto imu_raw = _m_imu_raw.get_ro();
+
+        // Current state
+        const auto curr_time = imu_raw->imu_time;
+        // TODO: should this be the last w or the estimated w in the VIO/integrator state?
+        const auto curr_w    = 0.5 * (imu_raw->w_hat + imu_raw->w_hat2);
+        const auto curr_quat = imu_raw->quat;
+        const auto curr_R    = quat_2_Rot(Eigen::Matrix<double, 4, 1>{curr_quat.x(), curr_quat.y(), curr_quat.z(), curr_quat.w()});
+        const auto curr_pos  = imu_raw->pos;
+        const auto curr_vel  = imu_raw->vel;
+
+        // Compute new quaternion
+
+        // Option 1: PG's code from issue 92. Same as Propagator.cpp
+        const auto future_q = rot_2_quat(exp_so3(-curr_w * dt) * curr_R);
+
+        // Option 2: PG's formula from issue 92. Same as Trawny eq. 96.
+        // const auto future_q = rot_2_quat(exp_so3(0.5 * Omega(curr_w) * dt) * curr_R);
+
+        // Compute new position
+        const auto delta_pos = curr_vel * dt;
+        const auto future_pos = curr_pos + delta_pos;
+
+        // Assemble state
+        Eigen::Matrix<double, 13, 1> state_plus = Eigen::Matrix<double, 13, 1>::Zero();
+        state_plus.block(0, 0, 4, 1) = future_q;
+        state_plus.block(4, 0, 3, 1) = future_pos;
+        state_plus.block(7, 0, 3, 1) = curr_vel;
+
+        return {state_plus, curr_time};
+    }
+
+    /**
+     * @brief Integrated quaternion from angular velocity
+     *
+     * See equation (48) of trawny tech report [Indirect Kalman Filter for 3D Attitude Estimation](http://mars.cs.umn.edu/tr/reports/Trawny05b.pdf).
+     *
+     */
+    static const inline Eigen::Matrix<double, 4, 4> Omega(Eigen::Matrix<double, 3, 1> w) {
+        Eigen::Matrix<double, 4, 4> mat;
+        mat.block(0, 0, 3, 3) = -skew_x(w);
+        mat.block(3, 0, 1, 3) = -w.transpose();
+        mat.block(0, 3, 3, 1) = w;
+        mat(3, 3) = 0;
+        return mat;
+    }
+
+    /**
+     * @brief Skew-symmetric matrix from a given 3x1 vector
+     *
+     * This is based on equation 6 in [Indirect Kalman Filter for 3D Attitude Estimation](http://mars.cs.umn.edu/tr/reports/Trawny05b.pdf):
+     * \f{align*}{
+     *  \lfloor\mathbf{v}\times\rfloor =
+     *  \begin{bmatrix}
+     *  0 & -v_3 & v_2 \\ v_3 & 0 & -v_1 \\ -v_2 & v_1 & 0
+     *  \end{bmatrix}
+     * @f}
+     *
+     * @param[in] w 3x1 vector to be made a skew-symmetric
+     * @return 3x3 skew-symmetric matrix
+     */
+    static const inline Eigen::Matrix<double, 3, 3> skew_x(const Eigen::Matrix<double, 3, 1> &w) {
```

Comment:
we can move these helper functions into one file. Duplicate across both predictors

Reply:
Not only both predictors but also rk4_integrator and pose_lookup.
Agree to make a unified file for shared util functions.